Remove pull from reconciler

chrfalch opened the pull request:
The current version of the reconciler uses a pull model when rendering, where each property is "materialised" on each render. This slows down the reconciler quite a bit, so a new model is proposed where the reconciler delegates property updates to its nodes and each node handles property updates by listening to changes in the Skia values through the dependency manager.

In addition we have a few places in the reconciler where we create new Skia objects on each render which is not very efficient since it invalidates the caching possibilities in the JSI implementations of the Skia objects (JsiHostObject).

*Updates:*
- JsiSkPaint (C++/web) has now got a `reset` method so that we can reuse paint objects without recreating them. 
- Canvas.tsx now has a few optimisations where we try to avoid recreating Skia objects and instead reuse existing ones.
- DependencyManager now tries to subscribe to values used in the component in an optimised way. Each value has only ONE subscription to the native value - and then each property in each component that uses the value subscribes to the DependencyManager instead. In addition some names have been refactored for a better fit to the implementation.
- Node/Declaration/Drawing has been updated to use a subscription model in stead of the pull model

Should be merged after #704

wcandillon commented on package/src/renderer/Canvas.tsx:
```diff
@@ -92,7 +93,7 @@ export const Canvas = forwardRef<SkiaView, CanvasProps>(
           canvasCtx.size.current = { width, height };
         }
-        const paint = Skia.Paint();
+        paint.reset();
```
👍🏻👍🏻👍🏻

chrfalch replied:
:)

wcandillon commented on package/src/renderer/Canvas.tsx:
```diff
@@ -101,7 +102,7 @@ export const Canvas = forwardRef<SkiaView, CanvasProps>(
           opacity: 1,
           ref,
-          center: Skia.Point(width / 2, height / 2),
+          center: { x: width / 2, y: height / 2 },
```
Is this change correct? If yes, can you elaborate?

chrfalch replied:
It was to avoid having to create a new Skia Point on every render.

wcandillon commented on package/src/renderer/nodes/Drawing.tsx:
```diff
@@ -47,20 +46,19 @@ export class DrawingNode<P> extends Node<P> {
   }
 
   draw(ctx: DrawingContext) {
-    const drawingProps = materialize(this.props);
```
isn't the issue in this function that `Node._props` is now `P` and not `Animated<P>`? This should enable the removing the `as P` references.

chrfalch replied:
The reasoning behind this was to keep the pattern simple, so when you set props, you are allowed to set both regular and animated versions of each prop (where Animated version is either a SkiaValue of the same type or a SkiaSelector). Then the props are resolved to values immediately (and on subsequent updates) and listeners are installed to update props whenever a SkiaValue change. 

Having the props directly mutated also ensures that we don't use double amount of memory for props in a Node.

chrfalch replied:
I see that this looks a bit wrong on the type side - we shouldn't need to cast the value like this, but it makes it a bit more efficient since we don't need to duplicate props.

wcandillon commented:
This is an exciting PR. Can't wait to ship this major improvement.
The `Paint.reset()` is great too and would be ready to be merged in another PR immediately.

Here are some comments: 
* The Node component should have an extra property named `materialized` that contain the property values.  Two things to consider here:
  * We may need the materialize function to get the "default" values (not sure if these are actually needed beyond to make TypeScript happy).
  * We may have a better name here (`_props` and `animatedProps`) but because this is quite fragile to refactor due to our use of JS setters/getters, we should probably move carefully here and rename only when we know the feature is working flawlessly
  * SubscriptionInfo` could look like this:
```jsx
type AnimatedProp = {
  name: string;
  value: SkiaValue<number>;
  unsubscribe: null | Unsubscribe;
};
type SubscriptionInfo = {
   // not in love with the props name here we could keep values as a name
  props: AnimatedProp[];
  unsubscribe: null | Unsubscribe;
};
```
This would make `subscribeNode` look like this:
```jsx
  subscribeNode(node: Node, props: Props) {
    const values = Object.keys(props)
      .map((name) => {
        const propValue = props[name];
        let unsubscribe: Unsubscribe | null = null;
        let value: SkiaValue<unknown>;
        if (isSelector(propValue)) {
          value = propValue.value;
          unsubscribe = value.addListener((v) => {
            node.materialized[name] = propValue.selector(v);
          });
        } else if (isValue(propValue)) {
          value = propValue;
          unsubscribe = value.addListener((v) => {
            node.materialized[name] = v;
          });
        } else {
          return null;
        }
        return { name, value, unsubscribe } as AnimatedProp;
      })
      .filter((v): v is AnimatedProp => v !== null);

    if (values.length > 0) {
      this.subscriptions.set(node, { props: values, unsubscribe: null });
    }
  }
```

And

```jsx
  unsubscribe() {
    this.subscriptions.forEach(({ unsubscribe, props }) => {
      props.forEach(({unsubscribe: unsub}) => {
        if (unsub) {
          unsub();
        } else {
          throw new Error("This shouldn't have happened 🤔")
        }
      })
      if (unsubscribe) {
        unsubscribe();
      }
    });
    this.subscriptions.clear();
  }
```

Does this look like a good model to do it? I'm interested to see if there are any oversights there.

wcandillon commented:
All the design goals you stated above are great and I would like to add one to the list: Nodes should know nothing about animations. It looks like everything could be setup in the DependencyManager. Or at least we should explore this objective.

Updating the node props directly is fantastic, but still, the types should be refined upstream. This is also where tests are really fun, we can write a test to check that the props object in the renderer is already a clone from the element props.

To contain everything in the dependency manager, I was thinking about the following data structure:

```jsx
type Mutator = (v: unknown) => void;
type SubscriptionState = {
  nodes: Map<Node, Mutator[]>;
  unsubscribeListener: null | Unsubscribe;
  // This one could be useful because it would allow us to remove subscriptions property
  //unsubscribeValue: null | Unsubscribe;
};
```


That would make the subscribeNode look something like this:
```jsx
  subscribeNode(node: Node, props: Props) {
    const values: SkiaValue<unknown>[] = [];
    mapKeys(props).forEach((prop) => {
      const value = props[prop];
      if (isValue(value)) {
        values.push(value);
        const state = this.getValueSubscription(value);
        const nodeState = this.getValueNodeSubscription(state, node);
        nodeState.push((v) => {
          node.props[prop] = v;
        });
        node.props[prop] = value.current;
      } else if (isSelector(value)) {
        values.push(value.value);
        // Same dance as above
      }
    });

    if (values.length > 0) {
      this.subscriptions.set(node, { values, unsubscribe: null });
    }
  }

```

and `subscribe()`:

```
    for (const [value, state] of this.values) {
      state.unsubscribeListener = value.addListener((v) => {
        [...state.nodes.values()].forEach((mutators) => {
          mutators.forEach((mutator) => {
            mutator(v);
          });
        });
      });
    }
```

If we could design this change in a way where Node doesn't anything about animations, I would be really excited.

chrfalch commented:
> All the design goals you stated above are great and I would like to add one to the list: Nodes should know nothing about animations. It looks like everything could be setup in the DependencyManager. Or at least we should explore this objective.

I'm a bit unsure about this requirement/use case? I agree about testability, but are there others?

I'll take a look and see what I can do - I have a strong opinion about making the dependency functionality as simple as possible to avoid extra performance and memory overhead as mentioned earlier.

> I couldn't get a sense of what `unregisterRedrawRequests ` or `dirtyFlag` are doing. This is also where tests could be helpful to get a sense of the kind of problems these little things are helping with.

The dirty flag was just something we discussed as a feature that could be used in the reconciler, it can easily be removed, will fix this :)

Will also come up with a better name for the `unregisterRedrawRequests` :) I see it doesn't make a lot of sense.

wcandillon commented:
About the requirement/use-case. I find it quite important to have the feature self-contained in `DependencyManager`. If it is possible to have `HostConfig.ts` and `Node.ts` not having any modifications (beyond the type refinement on `_props` we discussed), that would be wonderful.

The reasons I'm attached to this goal, provided that it is achievable, are the following:
* Keeping the Node class free from any animation logic will enable further performance optimizations
* The feature will be easier to understand/maintain/debug if self contained in the dependency manager which in its current form has a clear and simple contract. Separation of concerns is always great
* I'm seeing a small simplification in the dependency manager if we have the implementation self-contained, as mentioned in my latest review.

chrfalch commented:
Done!! :) 👍 
- moved subscription handling into dependency manager
- added unit tests for the dependency manager
- Cleaned up types for Node props.

wcandillon commented on package/src/renderer/HostConfig.ts:
```diff
@@ -101,11 +101,11 @@ const removeNode = (parent: Node, child: Node) => {
-  child.depMgr.unSubscribeNode(child);
   // unsubscribe to all children as well
   for (const c of child.children) {
-    removeNode(child, c);
```
The change here doesn't look correct, can you adjust it? I actually looks like this file should have 0 diffs, could that be?

wcandillon replied:
we remove `node.removeNode()` but `removeNode()` stays. I think this equals 0 diffs unless I'm mistaken.
If we take a step back, it makes sense that this file wouldn't have a diff?

chrfalch replied:
No, there was a mistake in the reconciler where it didn't remove child nodes - which is why this was added.

wcandillon commented on package/src/renderer/DependencyManager.tsx:
```diff
@@ -1,62 +1,200 @@
   }
 }
+
+const initializePropertySubscriptions = <P extends Record<string, unknown>>(
```
Should be:
```
const initializePropertySubscriptions = <P,>(
  node: Node<P>,
  props: AnimatedProps<P>
) => {
```

wcandillon commented on package/src/renderer/DependencyManager.tsx:
```diff
@@ -1,62 +1,200 @@
+  const nodePropSubscriptions: Array<{
+    value: SkiaValue<unknown>;
+    mutator: Mutator;
+    key: keyof P;
```
key can be removed now

chrfalch replied:
Ha ha, my debug code :) Removed

wcandillon reviewed (approved):
🥳
